File: internal/core/wiki/doc/schema.go
package doc

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"sort"
)

// schemaJSON is the shared node/mark vocabulary. It is a data file rather than
// a Go slice so the editor's TypeScript schema can be checked against the same
// bytes; see the comment inside it, and web/src/lib/codex/schema.test.ts.
//
//go:embed schema.json
var schemaJSON []byte

// The preservation placeholder types. Named constants because the capture path,
// the restore path, the projection and the editor all have to agree on them,
// and a typo in any one of them would look like ordinary unknown content.
const (
	// NodeUnknownContent is the block-level placeholder (ADR-0012 section 1).
	NodeUnknownContent = "unknownContent"

	// NodeUnknownInline is the inline placeholder.
	NodeUnknownInline = "unknownInline"

	// MarkUnknownMark is the mark placeholder.
	MarkUnknownMark = "unknownMark"
)

// The document vocabulary this phase added, named as constants for the same
// reason the preservation types are: the input rule, the projection, the
// publish-time tag aggregation and the editor all have to agree on them.
const (
	// NodeInlineTag is the inline `#tag` token. Its label is the text the
	// author typed; the slug it resolves to is derived server-side, so a tag
	// rename never has to rewrite a stored document.
	NodeInlineTag = "inlineTag"

	// MarkLink is the link mark, which carries all three of a link's possible
	// destinations — see [AttrLinkTargetTitle].
	MarkLink = "link"
)

// Attribute names the projection and the publish path read by name.
const (
	// AttrTagLabel is the text of an inline tag, as the author typed it.
	AttrTagLabel = "label"

	// AttrLinkHref is an external link's destination.
	AttrLinkHref = "href"

	// AttrLinkPageID is a resolved internal link's target page.
	AttrLinkPageID = "page_id"

	// AttrLinkTargetTitle is an UNRESOLVED wikilink's target: the title the
	// author wrote inside [[...]] when no page of that name was chosen.
	//
	// It is a third state, not a variant of the other two, and the distinction
	// is load-bearing. A link with a page_id resolves to a page; a link with an
	// href leaves Azimuthal; a link with only a target_title names a page that
	// does not exist yet, and clicking it offers to create one. Storing an
	// unresolved link as an href (say "#design-docs") would make it an external
	// link that goes nowhere, and storing it as a page_id is impossible —
	// there is no page.
	AttrLinkTargetTitle = "target_title"
)

// schema is the parsed manifest: type name -> group, plus the two position
// lists capture needs to choose between the block and inline placeholders, plus
// the attribute names the markdown projection reads.
type schema struct {
	Nodes              map[string]string `json:"nodes"`
	Marks              map[string]string `json:"marks"`
	InlineNodes        []string          `json:"inlineNodes"`
	InlineContentNodes []string          `json:"inlineContentNodes"`
	ProjectedAttrs     projectedAttrs    `json:"projectedAttrs"`
}

// projectedAttrs names the attributes [ToMarkdown] reads, per type. See the
// manifest's own comment for why an attribute rename fails differently from a
// type rename — nothing breaks, the content just stops being findable.
type projectedAttrs struct {
	Nodes map[string][]string `json:"nodes"`
	Marks map[string][]string `json:"marks"`
}

var parsedSchema = mustParseSchema()

func mustParseSchema() schema {
	var s schema
	if err := json.Unmarshal(schemaJSON, &s); err != nil {
		panic(fmt.Sprintf("wiki/doc: schema.json is not valid JSON: %v", err))
	}
	// The manifest is the boundary that decides what gets preserved. If it were
	// ever shipped without the placeholder types, capture would produce nodes
	// the next Shield treated as unknown again, wrapping them in an endless
	// nest. Fail at init rather than at runtime.
	for _, required := range []string{NodeUnknownContent, NodeUnknownInline} {
		if _, ok := s.Nodes[required]; !ok {
			panic(fmt.Sprintf("wiki/doc: schema.json omits the %q node, so preserved content would be re-captured on every save", required))
		}
	}
	if _, ok := s.Marks[MarkUnknownMark]; !ok {
		panic(fmt.Sprintf("wiki/doc: schema.json omits the %q mark", MarkUnknownMark))
	}
	// A projected attribute on a type that is not in the vocabulary describes a
	// projection that can never run. That is a manifest mistake rather than a
	// runtime state, so it fails at init like the missing placeholders above.
	for nodeType := range s.ProjectedAttrs.Nodes {
		if _, ok := s.Nodes[nodeType]; !ok {
			panic(fmt.Sprintf("wiki/doc: schema.json projects attributes of %q, which is not a node in the vocabulary", nodeType))
		}
	}
	for markType := range s.ProjectedAttrs.Marks {
		if _, ok := s.Marks[markType]; !ok {
			panic(fmt.Sprintf("wiki/doc: schema.json projects attributes of %q, which is not a mark in the vocabulary", markType))
		}
	}
	return s
}

// KnownNode reports whether the editor's schema defines this node type. A false
// answer means the node must be preserved rather than parsed.
func KnownNode(nodeType string) bool {
	_, ok := parsedSchema.Nodes[nodeType]
	return ok
}

// KnownMark reports whether the editor's schema defines this mark type.
func KnownMark(markType string) bool {
	_, ok := parsedSchema.Marks[markType]
	return ok
}

// SchemaNodes returns every node type the editor's schema defines, sorted.
func SchemaNodes() []string { return sortedKeys(parsedSchema.Nodes) }

// SchemaMarks returns every mark type the editor's schema defines, sorted.
func SchemaMarks() []string { return sortedKeys(parsedSchema.Marks) }

// SchemaJSON returns the raw manifest bytes, for the endpoint that lets the
// frontend assert it is registering the same vocabulary the server is
// preserving against.
func SchemaJSON() []byte { return schemaJSON }

// InlineNodes returns the node types that are themselves inline, sorted.
func InlineNodes() []string { return sortedCopy(parsedSchema.InlineNodes) }

// InlineContentNodes returns the node types whose children are inline, sorted.
func InlineContentNodes() []string { return sortedCopy(parsedSchema.InlineContentNodes) }

// ProjectedNodeAttrs returns the attribute names the markdown projection reads
// from each node type, sorted within each type.
func ProjectedNodeAttrs() map[string][]string { return copyAttrMap(parsedSchema.ProjectedAttrs.Nodes) }

// ProjectedMarkAttrs returns the attribute names the markdown projection reads
// from each mark type, sorted within each type.
func ProjectedMarkAttrs() map[string][]string { return copyAttrMap(parsedSchema.ProjectedAttrs.Marks) }

func copyAttrMap(in map[string][]string) map[string][]string {
	out := make(map[string][]string, len(in))
	for key, values := range in {
		out[key] = sortedCopy(values)
	}
	return out
}

// isInlineNode reports whether a node type is inline content.
func isInlineNode(nodeType string) bool {
	for _, n := range parsedSchema.InlineNodes {
		if n == nodeType {
			return true
		}
	}
	return false
}

// hasInlineContent reports whether a node type's children are inline content.
func hasInlineContent(nodeType string) bool {
	for _, n := range parsedSchema.InlineContentNodes {
		if n == nodeType {
			return true
		}
	}
	return false
}

func sortedCopy(in []string) []string {
	out := make([]string, len(in))
	copy(out, in)
	sort.Strings(out)
	return out
}

func sortedKeys(m map[string]string) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

File: web/src/lib/codex/schema.ts
/**
 * The Codex document schema vocabulary — the editor's half of the list the Go
 * layer preserves against.
 *
 * This file MIRRORS `internal/core/wiki/doc/schema.json`, and `schema.test.ts`
 * fails on any difference in either direction. It is a hand-written mirror
 * rather than an import because Vite cannot bundle a file from outside `web/`,
 * while a test running in Node can read it perfectly well — so the check is a
 * test rather than a build step.
 *
 * Why the drift matters, and why it is not symmetric (ADR-0012):
 *
 * - A type the editor registers but Go omits is merely captured when it did not
 *   need to be. The author sees an inert block where they could have had a real
 *   one. Annoying, and safe.
 * - A type Go omits from capture because it believes the editor handles it, when
 *   the editor does not, is **silent data loss**. ProseMirror drops content
 *   outside its schema without a word, and the drop happens on load, before
 *   anything server-side can notice.
 *
 * So: every name here must be registered by the editor, and every name the
 * editor registers must be here.
 */

/** Node types the editor's schema defines. */
export const CODEX_NODES = [
  // Core rich text (issue #15 E2).
  'doc',
  'paragraph',
  'text',
  'hardBreak',
  'heading',
  'bulletList',
  'orderedList',
  'listItem',
  'taskList',
  'taskItem',
  'blockquote',
  'codeBlock',
  'horizontalRule',
  'table',
  'tableRow',
  'tableHeader',
  'tableCell',
  'image',

  // ADR-0012 section 4 first-class macros.
  'panel',
  'expand',
  'statusLozenge',
  'tableOfContents',
  'layout',
  'layoutColumn',
  'childrenDisplay',
  'pageInclude',

  // The inline `#tag` token. Core rather than macro: it is ordinary authored
  // content, not an imported Confluence construct.
  'inlineTag',

  // ADR-0012's zero-silent-data-loss machinery.
  'unknownContent',
  'unknownInline',
] as const;

/** Mark types the editor's schema defines. */
export const CODEX_MARKS = [
  'bold',
  'italic',
  'strike',
  'code',
  'link',

  'unknownMark',
] as const;

/**
 * Node types that are themselves inline, and node types whose children are
 * inline.
 *
 * These exist so the server can choose between the block and the inline
 * preservation placeholder for a type it has never seen — it decides from the
 * position, because it cannot ask the type. Getting one wrong is not silent: a
 * placeholder that lands where ProseMirror will not accept it is dropped, the
 * server sees the id go missing, and publish refuses the write.
 */
export const CODEX_INLINE_NODES = [
  'text',
  'hardBreak',
  'statusLozenge',
  'inlineTag',
  'unknownInline',
] as const;

export const CODEX_INLINE_CONTENT_NODES = ['paragraph', 'heading', 'codeBlock'] as const;

export type CodexNodeName = (typeof CODEX_NODES)[number];
export type CodexMarkName = (typeof CODEX_MARKS)[number];

/** The three preservation carriers, named so nothing has to spell them twice. */
export const NODE_UNKNOWN_CONTENT = 'unknownContent';
export const NODE_UNKNOWN_INLINE = 'unknownInline';
export const MARK_UNKNOWN_MARK = 'unknownMark';

/** The inline tag token, and the link mark, named for the same reason. */
export const NODE_INLINE_TAG = 'inlineTag';
export const MARK_LINK = 'link';

/**
 * The attributes the server's markdown projection reads by name.
 *
 * Mirrors `projectedAttrs` in `internal/core/wiki/doc/schema.json`, and
 * `schema.test.ts` fails on any difference in either direction — the same
 * contract the type lists carry, guarding a different failure.
 *
 * A TYPE that drifts fails loudly in one direction and destroys content in the
 * other. An ATTRIBUTE that drifts does neither: the page publishes, the document
 * stores correctly, and the content quietly stops being findable, because
 * `internal/core/wiki/doc/text.go` looked for a name that is no longer there.
 * Nothing announces it. That is why the list is data checked by a test rather
 * than a comment asking people to remember.
 *
 * `extensions.test.ts` then asserts the real ProseMirror schema declares every
 * attribute named here, which is the half a list on its own cannot promise.
 */
export const PROJECTED_ATTRS = {
  nodes: {
    heading: ['level'],
    codeBlock: ['language'],
    image: ['attachment_id'],
    panel: ['kind'],
    expand: ['title'],
    statusLozenge: ['text'],
    pageInclude: ['page_id'],
    inlineTag: ['label'],
  },
  marks: {
    link: ['href', 'page_id', 'target_title'],
  },
} as const;

/**
 * The attribute names on the link mark, and the three states a link can be in.
 *
 * They are states, not variants — see `AttrLinkTargetTitle` in
 * `internal/core/wiki/doc/schema.go`:
 *
 * - `href` set: an external link, which leaves Azimuthal.
 * - `page_id` set: a resolved internal link. Never a URL, because a page's URL
 *   depends on the space it is being read in.
 * - `target_title` set: an UNRESOLVED wikilink — the author named a page that
 *   does not exist yet. Clicking it offers to create one.
 */
export const LINK_ATTRS = {
  href: 'href',
  pageId: 'page_id',
  targetTitle: 'target_title',
} as const;

/** The inline tag's one attribute: the label as the author typed it. */
export const TAG_ATTRS = { label: 'label' } as const;

/**
 * Attribute names on a preservation placeholder. Prefixed so they can never
 * collide with an attribute of the original node they describe.
 *
 * `az_raw` is a display copy: the server resolves the real original from its own
 * captured map and ignores whatever comes back here, so altering it changes
 * nothing about what gets stored. It is carried so the editor can label and size
 * the block, and so a reader can be shown what is being preserved.
 */
export const PRESERVED_ATTRS = {
  id: 'az_id',
  name: 'az_name',
  source: 'az_source',
  raw: 'az_raw',
  text: 'az_text',
} as const;

/** A ProseMirror document as it crosses the wire. */
export interface CodexDoc {
  type: 'doc';
  content?: CodexNode[];
}

export interface CodexNode {
  type: string;
  attrs?: Record<string, unknown>;
  content?: CodexNode[];
  marks?: CodexMark[];
  text?: string;
}

export interface CodexMark {
  type: string;
  attrs?: Record<string, unknown>;
}

/** The empty document — what a page with no content holds. */
export function emptyCodexDoc(): CodexDoc {
  return { type: 'doc', content: [] };
}

/**
 * Collects the preservation ids in a document, in document order.
 *
 * The editor takes this snapshot from the SERVER'S payload, before ProseMirror
 * parses it — which is the whole point. If ProseMirror drops a placeholder on
 * load, the id is in this snapshot and absent from the editor's state, so the
 * loss is detectable rather than invisible. Publish sends the difference as
 * `acknowledged_lost_ids` only for blocks the author actually deleted; anything
 * else missing is refused by the server.
 */
export function preservedIdsIn(node: CodexNode | CodexDoc | null | undefined): string[] {
  const ids: string[] = [];
  walk(node, (n) => {
    if (
      n.type === NODE_UNKNOWN_CONTENT ||
      n.type === NODE_UNKNOWN_INLINE ||
      n.type === MARK_UNKNOWN_MARK
    ) {
      const id = n.attrs?.[PRESERVED_ATTRS.id];
      if (typeof id === 'string' && id !== '') ids.push(id);
    }
  });
  return ids;
}

function walk(
  node: CodexNode | CodexDoc | null | undefined,
  visit: (n: CodexNode) => void,
): void {
  if (!node || typeof node !== 'object') return;
  const n = node as CodexNode;
  visit(n);
  n.marks?.forEach((mark) => walk(mark as CodexNode, visit));
  n.content?.forEach((child) => walk(child, visit));
}
